java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -354,10 +354,18 @@ private boolean isMethodParameter(Symbol symbol) {
     }
 
     private List<JavaFileScannerContext.Location> flowFromLearnedConstraints(ExplodedGraph.Edge edge, Set<LearnedConstraint> learnedConstraints) {
-      return learnedConstraints.stream()
+      // FIXME SONARJAVA-2303 temporary workaround to avoid duplicated messages when constraint is reported on relational SV and also on its operand
+      List<JavaFileScannerContext.Location> result = new ArrayList<>();
+      Set<String> messages = new HashSet<>();
+      learnedConstraints.stream()
         .map(lc -> learnedConstraintFlow(lc, edge))
         .flatMap(List::stream)
-        .collect(Collectors.toList());
+        .forEach(location -> {
+          if (!location.syntaxNode.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO) || messages.add(location.msg)) {
```

Comment:
Mmmh, I'm not sure to follow you here. What is bothering me is having two times the same message on the same tree (so a duplicate of `Location`). So why considering the kind of the tree and only the message value? It's probably way simpler to rely on equality of locations to filter out, with a `LinkedHashSet` for instance.